datastore/mqtt/mqtt.go

```diff
@@ -0,0 +1,246 @@
+	"github.com/teslamotors/fleet-telemetry/telemetry"
+)
+
+type MQTTProducer struct {
+	client             pahomqtt.Client
+	config             *Config
+	logger             *logrus.Logger
+	airbrakeHandler    *airbrake.AirbrakeHandler
+	namespace          string
+	ctx                context.Context
+	ackChan            chan (*telemetry.Record)
+	reliableAckTxTypes map[string]interface{}
+}
+
+type Config struct {
+	Broker               string `json:"broker"`
+	ClientID             string `json:"client_id"`
+	Username             string `json:"username"`
+	Password             string `json:"password"`
+	TopicBase            string `json:"topic_base"`
+	QoS                  byte   `json:"qos"`
+	Retained             bool   `json:"retained"`
+	ConnectTimeout       int    `json:"connect_timeout_ms"`
+	PublishTimeout       int    `json:"publish_timeout_ms"`
+	DisconnectTimeout    int    `json:"disconnect_timeout_ms"`
+	ConnectRetryInterval int    `json:"connect_retry_interval_ms"`
+	KeepAlive            int    `json:"keep_alive_seconds"`
+}
+
+type Metrics struct {
+	errorCount       adapter.Counter
+	publishCount     adapter.Counter
+	byteTotal        adapter.Counter
+	reliableAckCount adapter.Counter
+}
+
+const (
+	DefaultPublishTimeout       = 2500
+	DefaultConnectTimeout       = 30000
+	DefaultConnectRetryInterval = 10000
+	DefaultDisconnectTimeout    = 250
+	DefaultKeepAlive            = 30
+	DefaultQoS                  = 0
+)
+
+var (
+	metricsRegistry Metrics
+	metricsOnce     sync.Once
+)
+
+// Allow us to mock the mqtt.NewClient function for testing
+var PahoNewClient = pahomqtt.NewClient
+
+func NewProducer(ctx context.Context, config *Config, metrics metrics.MetricCollector, namespace string, airbrakeHandler *airbrake.AirbrakeHandler, ackChan chan (*telemetry.Record), reliableAckTxTypes map[string]interface{}, logger *logrus.Logger) (telemetry.Producer, error) {
+	registerMetricsOnce(metrics)
+
+	// Set default values
+	if config.PublishTimeout == 0 {
+		config.PublishTimeout = DefaultPublishTimeout
+	}
+	if config.ConnectTimeout == 0 {
+		config.ConnectTimeout = DefaultConnectTimeout
+	}
+	if config.DisconnectTimeout == 0 {
+		config.DisconnectTimeout = DefaultDisconnectTimeout
+	}
+	if config.ConnectRetryInterval == 0 {
+		config.ConnectRetryInterval = DefaultConnectRetryInterval
+	}
+	if config.QoS == 0 {
+		config.QoS = DefaultQoS
+	}
+	if config.KeepAlive == 0 {
+		config.KeepAlive = DefaultKeepAlive
+	}
+
+	opts := pahomqtt.NewClientOptions().
+		AddBroker(config.Broker).
+		SetClientID(config.ClientID).
+		SetUsername(config.Username).
+		SetPassword(config.Password).
+		SetConnectRetry(true).
+		SetAutoReconnect(true).
+		SetConnectRetryInterval(time.Duration(config.ConnectRetryInterval) * time.Millisecond).
+		SetConnectTimeout(time.Duration(config.ConnectTimeout) * time.Millisecond).
+		SetOrderMatters(false).
+		SetKeepAlive(time.Duration(config.KeepAlive) * time.Second)
+
+	client := PahoNewClient(opts)
+	client.Connect()
```

Comment:
ConnectRetry is set. So, the client.Connect() might fail, but that's not a problem because the Paho client will keep trying to connect.
This will make sure a telemetry container can start even if a configured MQTT broker is not online.